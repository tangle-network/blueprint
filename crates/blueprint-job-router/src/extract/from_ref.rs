/// Used to do reference-to-value conversions thus not consuming the input value.
///
/// This is mainly used with [`Context`] to extract "subcontexts" from a reference to main application
/// state.
///
/// See [`Context`] for more details on how library authors should use this trait.
///
/// This trait can be derived using `#[derive(FromRef)]`.
///
/// [`Context`]: https://docs.rs/axum/0.8/axum/extract/struct.Context.html
// NOTE: This trait is defined in axum-core, even though it is mainly used with `Context` which is
// defined in axum. That allows crate authors to use it when implementing extractors.
pub trait FromRef<T> {
    /// Converts to this type from a reference to the input type.
    fn from_ref(input: &T) -> Self;
}

impl<T> FromRef<T> for T
where
    T: Clone,
{
    fn from_ref(input: &T) -> Self {
        input.clone()
    }
}
