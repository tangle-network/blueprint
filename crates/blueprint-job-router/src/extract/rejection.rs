//! Rejection response types.

use crate::__define_rejection as define_rejection;

define_rejection! {
    #[body = "Request body didn't contain valid UTF-8"]
    /// Rejection type used when buffering the request into a [`String`] if the
    /// body doesn't contain valid UTF-8.
    pub struct InvalidUtf8(Error);
}
