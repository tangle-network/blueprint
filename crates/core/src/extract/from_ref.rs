/// Used to do reference-to-value conversions thus not consuming the input value.
///
/// This is mainly used with [`Context`] to extract "subcontexts" from a reference to main application
/// state.
///
/// See [`Context`] for more details on how library authors should use this trait.
///
/// This trait can be derived using `#[derive(FromRef)]`.
///
/// [`Context`]: https://docs.rs/blueprint_sdk/latest/blueprint_sdk/extract/struct.Context.html
pub trait FromRef<T> {
    /// Converts to this type from a reference to the input type.
    fn from_ref(input: &T) -> Self;
}

impl<T> FromRef<T> for T
where
    T: Clone,
{
    fn from_ref(input: &T) -> Self {
        input.clone()
    }
}
