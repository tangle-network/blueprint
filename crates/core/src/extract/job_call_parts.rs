use super::{FromJobCall, FromJobCallParts, rejection::*};
use crate::JobCall;
use crate::job_call::Parts;
use crate::metadata::{MetadataMap, MetadataValue};
use alloc::string::String;
use bytes::{Bytes, BytesMut};
use core::convert::Infallible;

impl<Ctx> FromJobCall<Ctx> for JobCall
where
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call(call: JobCall, _: &Ctx) -> Result<Self, Self::Rejection> {
        Ok(call)
    }
}

impl<Ctx> FromJobCallParts<Ctx> for MetadataMap<MetadataValue>
where
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call_parts(parts: &mut Parts, _: &Ctx) -> Result<Self, Self::Rejection> {
        let metadata = parts.metadata.clone();

        Ok(metadata)
    }
}

impl<Ctx> FromJobCall<Ctx> for BytesMut
where
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call(call: JobCall, _: &Ctx) -> Result<Self, Self::Rejection> {
        let (_, body) = call.into_parts();
        Ok(body.into())
    }
}

impl<Ctx> FromJobCall<Ctx> for Bytes
where
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call(call: JobCall, _: &Ctx) -> Result<Self, Self::Rejection> {
        let (_, body) = call.into_parts();

        Ok(body)
    }
}

impl<Ctx> FromJobCall<Ctx> for String
where
    Ctx: Send + Sync,
{
    type Rejection = InvalidUtf8;

    async fn from_job_call(call: JobCall, ctx: &Ctx) -> Result<Self, Self::Rejection> {
        let bytes = Bytes::from_job_call(call, ctx)
            .await
            .map_err(InvalidUtf8::from_err)?;

        let string = String::from_utf8(bytes.into()).map_err(InvalidUtf8::from_err)?;

        Ok(string)
    }
}

impl<Ctx> FromJobCallParts<Ctx> for Parts
where
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call_parts(
        parts: &mut Parts,
        _context: &Ctx,
    ) -> Result<Self, Self::Rejection> {
        Ok(parts.clone())
    }
}
