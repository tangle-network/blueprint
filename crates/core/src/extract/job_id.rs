//! Extracts the Job ID from the current job call.

use crate::job_call::Parts as JobCallParts;
use crate::{FromJobCallParts, IntoJobResult};

/// A Specialized extractor for the [`JobId`] of the current [`JobCall`].
///
/// It will try to extract `T` from the job call parts as long as `T` implements [`TryFrom`](core::convert::TryFrom) for [`JobId`] which
/// already works for all the primitive types and can be implemented for custom types.
///
/// [`JobId`]: crate::JobId
/// [`JobCall`]: crate::JobCall
#[derive(Debug, Clone)]
pub struct JobId<T>(pub T);

crate::__impl_deref!(JobId);

impl<T, Ctx> FromJobCallParts<Ctx> for JobId<T>
where
    T: TryFrom<crate::JobId>,
    <T as TryFrom<crate::JobId>>::Error: IntoJobResult,
    Ctx: Send + Sync + 'static,
{
    type Rejection = <T as TryFrom<crate::JobId>>::Error;

    async fn from_job_call_parts(
        parts: &mut JobCallParts,
        _: &Ctx,
    ) -> Result<Self, Self::Rejection> {
        let job_id = parts.job_id;
        T::try_from(job_id).map(JobId)
    }
}

#[cfg(test)]
mod tests {

    use crate::IntoJobId;

    use super::*;

    macro_rules! test_extract_job_id{
        ($($ty:ty),*) => {
            $(
                let job_id: $ty = 42;
                let mut parts = JobCallParts::new(job_id);
                let extracted_job_id: JobId<$ty> = JobId::from_job_call_parts(&mut parts, &()).await.unwrap();
                assert_eq!(extracted_job_id.0, job_id);

                let job_id = <$ty>::MAX;
                let mut parts = JobCallParts::new(job_id);
                let extracted_job_id: JobId<$ty> = JobId::from_job_call_parts(&mut parts, &()).await.unwrap();
                assert_eq!(extracted_job_id.0, job_id);

                let job_id = <$ty>::MIN;
                let mut parts = JobCallParts::new(job_id);
                let extracted_job_id: JobId<$ty> = JobId::from_job_call_parts(&mut parts, &()).await.unwrap();
                assert_eq!(extracted_job_id.0, job_id);
            )*
        };
    }

    #[tokio::test]
    async fn it_correctly_extracts_all_numbers() {
        test_extract_job_id!(
            u8, u16, u32, u64, usize, i8, i16, i32, i64, isize, i128, u128
        );
    }

    #[derive(Copy, Clone, Debug, PartialEq)]
    struct MyCustomJobId;

    impl TryFrom<crate::JobId> for MyCustomJobId {
        type Error = ();

        fn try_from(job_id: crate::JobId) -> Result<Self, Self::Error> {
            if u64::from(job_id) == 42u64 {
                Ok(MyCustomJobId)
            } else {
                Err(())
            }
        }
    }

    impl IntoJobId for MyCustomJobId {
        fn into_job_id(self) -> crate::JobId {
            42u64.into()
        }
    }

    #[tokio::test]
    async fn it_works_for_custom_types() {
        let job_id: MyCustomJobId = MyCustomJobId;
        let mut parts = JobCallParts::new(job_id);

        let extracted_job_id: JobId<MyCustomJobId> =
            JobId::from_job_call_parts(&mut parts, &()).await.unwrap();
        assert_eq!(extracted_job_id.0, job_id);
    }

    #[tokio::test]
    async fn it_rejects_invalid_job_ids() {
        let mut parts = JobCallParts::new(41u64);
        let extracted_job_id = JobId::<MyCustomJobId>::from_job_call_parts(&mut parts, &()).await;
        assert!(extracted_job_id.is_err());
    }
}
