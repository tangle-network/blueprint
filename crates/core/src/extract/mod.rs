//! Types and traits for extracting data from requests.
//!
//! See [`axum::extract`] for more details.
//!
//! [`axum::extract`]: https://docs.rs/axum/0.8/axum/extract/index.html

use crate::JobCall;
use crate::job_call::Parts;
use crate::job_result::IntoJobResult;
use core::convert::Infallible;
use core::future::Future;

pub mod rejection;

mod from_ref;
mod job_call_parts;
mod option;
mod tuple;

pub use self::{
    from_ref::FromRef,
    option::{OptionalFromJobCall, OptionalFromJobCallParts},
    rejection::InvalidUtf8,
};

mod private {
    #[derive(Debug, Clone, Copy)]
    pub enum ViaParts {}

    #[derive(Debug, Clone, Copy)]
    pub enum ViaJobCall {}
}

/// Types that can be created from job calls.
///
/// Extractors that implement `FromJobCall` can consume the job call body and can thus only be run
/// once for handlers.
///
/// If your extractor doesn't need to consume the job call body then you should implement
/// [`FromJobCallParts`] and not [`FromJobCall`].
///
/// See [`axum::extract`] for more general docs about extractors.
///
/// [`axum::extract`]: https://docs.rs/axum/0.8/axum/extract/index.html
#[diagnostic::on_unimplemented(
    note = "Function argument is not a valid blueprint extractor. \nSee `https://docs.rs/axum/0.8/axum/extract/index.html` for details"
)]
pub trait FromJobCallParts<Ctx>: Sized {
    /// If the extractor fails it'll use this "rejection" type. A rejection is
    /// a kind of error that can be converted into a job result.
    type Rejection: IntoJobResult;

    /// Perform the extraction.
    fn from_job_call_parts(
        parts: &mut Parts,
        ctx: &Ctx,
    ) -> impl Future<Output = Result<Self, Self::Rejection>> + Send;
}

/// Types that can be created from job calls.
///
/// Extractors that implement `FromJobCall` can consume the job call body and can thus only be run
/// once for handlers.
///
/// If your extractor doesn't need to consume the job call body then you should implement
/// [`FromJobCallParts`] and not [`FromJobCall`].
///
/// See [`axum::extract`] for more general docs about extractors.
///
/// [`axum::extract`]: https://docs.rs/axum/0.8/axum/extract/index.html
#[diagnostic::on_unimplemented(
    note = "Function argument is not a valid blueprint extractor. \nSee `https://docs.rs/axum/0.8/axum/extract/index.html` for details"
)]
pub trait FromJobCall<Ctx, M = private::ViaJobCall>: Sized {
    /// If the extractor fails it'll use this "rejection" type. A rejection is
    /// a kind of error that can be converted into a job result.
    type Rejection: IntoJobResult;

    /// Perform the extraction.
    fn from_job_call(
        call: JobCall,
        ctx: &Ctx,
    ) -> impl Future<Output = Result<Self, Self::Rejection>> + Send;
}

impl<Ctx, T> FromJobCall<Ctx, private::ViaParts> for T
where
    Ctx: Send + Sync,
    T: FromJobCallParts<Ctx>,
{
    type Rejection = <Self as FromJobCallParts<Ctx>>::Rejection;

    async fn from_job_call(call: JobCall, ctx: &Ctx) -> Result<Self, Self::Rejection> {
        let (mut parts, _) = call.into_parts();
        Self::from_job_call_parts(&mut parts, ctx).await
    }
}

impl<Ctx, T> FromJobCallParts<Ctx> for Result<T, T::Rejection>
where
    T: FromJobCallParts<Ctx>,
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call_parts(parts: &mut Parts, ctx: &Ctx) -> Result<Self, Self::Rejection> {
        Ok(T::from_job_call_parts(parts, ctx).await)
    }
}

impl<Ctx, T> FromJobCall<Ctx> for Result<T, T::Rejection>
where
    T: FromJobCall<Ctx>,
    Ctx: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call(call: JobCall, ctx: &Ctx) -> Result<Self, Self::Rejection> {
        Ok(T::from_job_call(call, ctx).await)
    }
}
