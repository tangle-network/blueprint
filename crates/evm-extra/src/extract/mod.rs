//! Extraction utilities for EVM-specific data
//!
//! This module provides extractors for EVM blocks and events, following the same
//! pattern as Tangle for consistency and reusability.

pub mod block;
pub mod event;

pub use block::{BlockHash, BlockNumber, BlockTimestamp};
pub use event::{BlockEvents, Events};
