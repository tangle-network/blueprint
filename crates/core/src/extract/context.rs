use crate::extract::{FromJobCallParts, FromRef};
use crate::job_call::Parts;
use core::{
    convert::Infallible,
    ops::{Deref, DerefMut},
};

/// Extractor for context.
///
/// See ["Accessing context in middleware"][context-from-middleware] for how to
/// access context in middleware.
///
/// context is global and used in every request a router with context receives.
/// For accessing data derived from requests, such as authorization data, see [`Extension`].
///
/// [context-from-middleware]: crate::middleware#accessing-context-in-middleware
/// [`Extension`]: crate::Extension
///
/// # With `Router`
///
/// ```
/// use blueprint_sdk::extract::Context;
/// use blueprint_sdk::{Job, Router};
///
/// // the application context
/// //
/// // here you can put configuration, database connection pools, or whatever
/// // context you need
/// #[derive(Clone)]
/// struct AppContext {}
///
/// let context = AppContext {};
///
/// const MY_JOB_ID: u32 = 0;
///
/// // create a `Router` that holds our context
/// let app = Router::new()
///     .route(MY_JOB_ID, handler)
///     // provide the context so the router can access it
///     .with_context(context);
///
/// async fn handler(
///     // access the context via the `Context` extractor
///     // extracting a context of the wrong type results in a compile error
///     Context(context): Context<AppContext>,
/// ) {
///     // use `context`...
/// }
/// # let _: Router = app;
/// ```
///
/// Note that `context` is an extractor, so be sure to put it before any body
/// extractors, see ["the order of extractors"][order-of-extractors].
///
/// [order-of-extractors]: crate::extract#the-order-of-extractors
///
/// # With `Job`
///
/// ```
/// use blueprint_sdk::Job;
/// use blueprint_sdk::extract::Context;
///
/// #[derive(Clone)]
/// struct AppContext {}
///
/// let context = AppContext {};
///
/// async fn job(Context(context): Context<AppContext>) {
///     // use `context`...
/// }
///
/// // provide the context so the job can access it
/// let job_with_context = job.with_context(context);
/// ```
///
/// # Substates
///
/// [`context`] only allows a single context type but you can use [`FromRef`] to extract "substates":
///
/// ```
/// use blueprint_sdk::Router;
/// use blueprint_sdk::extract::{Context, FromRef};
///
/// // the application context
/// #[derive(Clone)]
/// struct AppContext {
///     // that holds some api specific context
///     api_state: ApiContext,
/// }
///
/// // the api specific context
/// #[derive(Clone)]
/// struct ApiContext {}
///
/// // support converting an `AppContext` in an `ApiContext`
/// impl FromRef<AppContext> for ApiContext {
///     fn from_ref(app_state: &AppContext) -> ApiContext {
///         app_state.api_state.clone()
///     }
/// }
///
/// let context = AppContext {
///     api_state: ApiContext {},
/// };
///
/// const HANDLER_JOB_ID: u32 = 0;
/// const FETCH_API_JOB_ID: u32 = 1;
///
/// let app = Router::new()
///     .route(HANDLER_JOB_ID, handler)
///     .route(FETCH_API_JOB_ID, fetch_api)
///     .with_context(context);
///
/// async fn fetch_api(
///     // access the api specific context
///     Context(api_state): Context<ApiContext>,
/// ) {
/// }
///
/// async fn handler(
///     // we can still access to top level context
///     Context(context): Context<AppContext>,
/// ) {
/// }
/// # let _: Router = app;
/// ```
///
/// For convenience `FromRef` can also be derived using `#[derive(FromRef)]`.
///
/// # For library authors
///
/// If you're writing a library that has an extractor that needs context, this is the recommended way
/// to do it:
///
/// ```rust
/// use blueprint_sdk::extract::{FromJobCallParts, FromRef};
/// use blueprint_sdk::job_call::Parts;
/// use std::convert::Infallible;
///
/// // the extractor your library provides
/// struct MyLibraryExtractor;
///
/// impl<S> FromJobCallParts<S> for MyLibraryExtractor
/// where
///     // keep `S` generic but require that it can produce a `MyLibraryContext`
///     // this means users will have to implement `FromRef<UserContext> for MyLibraryContext`
///     MyLibraryContext: FromRef<S>,
///     S: Send + Sync,
/// {
///     type Rejection = Infallible;
///
///     async fn from_job_call_parts(
///         parts: &mut Parts,
///         context: &S,
///     ) -> Result<Self, Self::Rejection> {
///         // get a `MyLibraryContext` from a reference to the context
///         let context = MyLibraryContext::from_ref(context);
///
///         // ...
///         # todo!()
///     }
/// }
///
/// // the context your library needs
/// struct MyLibraryContext {
///     // ...
/// }
/// ```
///
/// # Shared mutable context
///
/// [As context is global within a `Router`][global] you can't directly get a mutable reference to
/// the context.
///
/// The most basic solution is to use an `Arc<Mutex<_>>`. Which kind of mutex you need depends on
/// your use case. See [the tokio docs] for more details.
///
/// Note that holding a locked `std::sync::Mutex` across `.await` points will result in `!Send`
/// futures which are incompatible with `blueprint_sdk`. If you need to hold a mutex across `.await` points,
/// consider using a `tokio::sync::Mutex` instead.
///
/// ## Example
///
/// ```
/// use blueprint_sdk::Router;
/// use blueprint_sdk::extract::Context;
/// use std::sync::{Arc, Mutex};
///
/// #[derive(Clone)]
/// struct AppContext {
///     data: Arc<Mutex<String>>,
/// }
///
/// async fn job(Context(context): Context<AppContext>) {
///     {
///         let mut data = context.data.lock().expect("mutex was poisoned");
///         *data = "updated foo".to_owned();
///     }
///
///     // ...
/// }
///
/// let context = AppContext {
///     data: Arc::new(Mutex::new("foo".to_owned())),
/// };
///
/// let app = Router::new().route("/", job).with_context(context);
/// # let _: Router = app;
/// ```
///
/// [global]: crate::Router::with_state
/// [the tokio docs]: https://docs.rs/tokio/1.25.0/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use
#[derive(Debug, Default, Clone, Copy)]
pub struct Context<S>(pub S);

impl<OuterContext, InnerContext> FromJobCallParts<OuterContext> for Context<InnerContext>
where
    InnerContext: FromRef<OuterContext>,
    OuterContext: Send + Sync,
{
    type Rejection = Infallible;

    async fn from_job_call_parts(
        _parts: &mut Parts,
        ctx: &OuterContext,
    ) -> Result<Self, Self::Rejection> {
        let inner_state = InnerContext::from_ref(ctx);
        Ok(Self(inner_state))
    }
}

impl<S> Deref for Context<S> {
    type Target = S;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<S> DerefMut for Context<S> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}
