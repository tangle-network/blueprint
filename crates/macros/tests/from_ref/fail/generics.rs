use blueprint_sdk::extract::FromRef;

#[derive(Clone, FromRef)]
struct AppContext<T> {
    foo: T,
}

fn main() {}
