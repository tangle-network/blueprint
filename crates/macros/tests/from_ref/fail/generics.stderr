error: `#[derive(FromRef)]` doesn't support generics
 --> tests/from_ref/fail/generics.rs:4:18
  |
4 | struct AppContext<T> {
  |                  ^^^
